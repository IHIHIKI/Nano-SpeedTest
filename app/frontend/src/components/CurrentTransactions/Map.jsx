import React, { Component } from 'react';
import { connect } from 'react-redux';


const base = 'https://maps.googleapis.com/maps/api/staticmap';
const scale = '4';
let sizeTemp = '1500x390';

const size = sizeTemp;

const googleKey = process.env.REACT_APP_GOOGLE_MAPS_KEY;

const makeCoordinate = (point) => {
    return `${point.latitude},${point.longitude}`;
};


class Map extends Component  {


    render() {
         let transactionPairs = [];
         let markerPairs = []
         const {numToRerun, table } = this.props;
         const mostRecent = table && table.length && table.slice(table.length - numToRerun); // get the last numToRerun transactions

         console.log(mostRecent)
         if(mostRecent.length > 0){
           console.log(mostRecent.length)
           mostRecent.forEach((trans) => {
                transactionPairs.push(`${makeCoordinate(trans.origin)}|${makeCoordinate(trans.destination)}`);
                markerPairs.push(`markers=color:0x4A90E2|${makeCoordinate(trans.origin)}&markers=color:0x4A90E2|${makeCoordinate(trans.destination)}`);
            });

            const paths = transactionPairs.join('&path='); // each needs to be draw separately
            const markers = markerPairs.join("&")

            return (
                <img className="transaction-box-map"
                src={`${base}?size=${size}&zoom=2&scale=${scale}&${markers}&path=${paths}&key=${googleKey}`}
                alt={'Google Map Showing Destination and Origin'} />
            );
         }
        return(<div className="loader"></div>)
    }

};

const mapStateToProps = (state) => {
    return {
        numToRerun: state.table.num,
        table: state.table.rows,
        initialValues: {
            origin: Object.keys(state.nodes).length ? state.nodes[Object.keys(state.nodes)[0]].id : undefined,
            destination: Object.keys(state.nodes).length ? state.nodes[Object.keys(state.nodes)[2]].id : undefined
        }
    };

};

export default connect(mapStateToProps)(Map);
