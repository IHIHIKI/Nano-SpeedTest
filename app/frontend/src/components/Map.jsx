import React from 'react';
import { compose, withProps } from 'recompose';
import { withGoogleMap, GoogleMap, Polyline, Marker } from 'react-google-maps';
import '../styles/GoogleMaps.css';

class TestMap extends React.Component {
    state = {
        mapLoaded: false
    };
    // Updates the map window once it is loaded to only show the distance between the two cities
    recomputeBounds() {
        const {cities} = this.props;
        const bounds = new window.google.maps.LatLngBounds();
        cities.map((city) => bounds.extend(city.coords));
        this.refs.map.fitBounds(bounds);
        this.setState({mapLoaded: true});
    }
    render() {
        const {cities} = this.props;
        const {mapLoaded} = this.state;

        return (
            <GoogleMap
                ref='map'
                defaultZoom={2}
                defaultCenter={{ lat: 0, lng: 0 }}
                defaultOptions={{ // disable moving and zooming the map
                    zoomControl: false,
                    gestureHandling: 'none',
                    draggableCursor: 'default',
                    disableDefaultUI: true
                }}
                onTilesLoaded={() => this.recomputeBounds()}
            >
            { // the map needs to render before recomputing and drawing markers and lines
                mapLoaded ? (
                    <div>
                        <Polyline
                            path={[this.props.cities[0].coords, this.props.cities[1].coords]}
                            options={{
                                clickable: false,
                                strokeOpacity: 1,
                                strokeWeight: 2,
                                geodesic: true,
                            }}
                        />
                        {cities.map((city) =>
                            <Marker
                                key={city.name}
                                title={city.name}
                                position={city.coords}
                                clickable={false}
                            />
                        )}
                    </div>

                ) : null

            }

            </GoogleMap>
        );
  }
}

const Map = compose(
    withProps({
        loadingElement: <div className={'GoogleMapsLoading'} />,
        containerElement: <div className={'GoogleMapsContainer'} />,
        mapElement: <div className={'GoogleMapsElement'} />
    }),
    withGoogleMap
)(TestMap);

export default Map;
