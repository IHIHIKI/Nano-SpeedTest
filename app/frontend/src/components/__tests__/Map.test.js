import React from 'react';
import renderer from 'react-test-renderer';
import Map from '../CurrentTransactions/Map';
import { shallow } from 'enzyme';

it('shallow renders without crashing', () => {
    const testMapData = [{
        origin: {
            latitude: "1.293100",
            longitude: "103.850100"
        },
        destination: {
            latitude: "40.804300",
            longitude: "-74.012100"
        }
    }];
    shallow(<Map transactions={testMapData}/>);
});

it(`renders snapshot correctly`, () => {
    const testMapData = [{
        origin: {
            latitude: "1.293100",
            longitude: "103.850100"
        },
        destination: {
            latitude: "40.804300",
            longitude: "-74.012100"
        }
    }];
    const tree = renderer.create(
        <Map transactions={testMapData}/>
    ).toJSON();
    expect(tree).toMatchSnapshot();
});
