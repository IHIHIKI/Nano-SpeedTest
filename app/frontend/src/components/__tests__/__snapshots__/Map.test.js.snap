// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot correctly 1`] = `
<img
  alt="Google Map Showing Destination and Origin"
  className="GoogleMap"
  src="https://maps.googleapis.com/maps/api/staticmap?size=1200x400&scale=2&markers=size:small|1.293100,103.850100|40.804300,-74.012100&path=1.293100,103.850100|40.804300,-74.012100&key=AIzaSyAovwwttuLBsvolc97QUc5pejsbwfLPPZs"
/>
`;
